python_packaging/roadmap.rst

```diff
@@ -184,16 +184,69 @@ sed 's/ -->/\)\*/'
+
+  - Some kind of nested structure likely appropriate, perhaps
+    structured as nested CMake projects that in principle could stand
+    alone. That's probably workable because nested projects can see
+    the parent project's cache variables (TODO check this)
+  - probably a top-level project coordinating a libgromacs build and a
+    python module build, with the former typically feeding the latter
+  - the libgromacs build may be able to leverage independent efforts
+    towards a multi-configuration build (so SIMD/MPI/GPU agnostic)
+  - top-level project offers much the same UI as now, passing much of
+    it through to the libgromacs project
+  - top-level project offers the option to find a Python (or be told
+    which to use), to find a libgromacs (or be told, or be told to
+    build), to build any necessary wrapper binaries (ie. classical gmx
+    and mdrun), and to deploy all linked artefacts to
+    CMAKE_INSTALL_PREFIX or the appropriate Python site-packages
+  - the top-level project will be used by e.g. setup.py wrapper
+    from scikit-build/distutils
+  - requires reform of compiler flags handling
+  - probably requires some re-organization of external dependencies
+    of libgromacs
+  - follow online "Modern CMake" best practices as far as practicable
+  - library should be available for static linking with position
+    independent code to allow a single shared object to be built for
+    the Python module.
+
+* Dissolve boundary between libgmxapi and libgromacs
+
+  - no effort on form and stability of the C++ headers and library in
+    2019, beyond what facilitates implementing the Python interface
+  - existing libgromacs declarations of "public API" and installed
+    headers removed
+
+* libgromacs to be able to be use an MPI communicator passed in,
+  rather than hard-coding MPI_COMM_WORLD anywhere. It is likely that
+  existing wrapper binaries can use the same mechanism to pass
+  MPI_COMM_WORLD to libgromacs.
+
+* UI helpers should express - TODO Eric, we need to clarify what is
+  meant here.
   - preferred name for datum as a string
   - setter
   - typing and type discovery
   - help text
   - (for CLI: short name for flag)
-* library should be available for static linking with position independent code
-  to allow a single shared-object to be built for the Python module.
+
+Possible GROMACS source changes whose impact is currently unknown
+=================================================================
+* gmx::Any (which is a flavour of C++17 std::any) type could be
+  helpful at API boundary. Also perhaps a flavour of C++17
+  std::optional or std::variant.
+
+GROMACS source changes deferred to later in 2019
+================================================
+* Build system works also from tarball
+* Build system can produce maximally static artefacts (for performance
+  on HPC infrastructure)
+* express grompp and mdrun options handling with gmx::Options to
```

Comment:
Dictionary-like handling is fine for UI, but for input checking, serialization and workflow checkpointing, API implementation, extension code, and the opportunity for compartmentalized optimization, we should also be able to get an expression of basic typing or hierarchical data structures. As successive features, I would think something like
1. Punt parameters from Python to C++ as a dictionary and be prepared for errors.
2. Allow Python to discover appropriate named parameters from implementation.
3. Provide enough information through API for input checking / conversion, etc. in the UI implementation.
4. Provide enough information in API headers to allow compile-time handling of typed inputs/outputs/parameters.

Somewhere in there, we should start building facilities to allow the separation of inputs/outputs/parameters from the storage format. E.g. A module can accept selections or atom index info -> UI implementation provides a future-like object for the info, possibly using a file reader for a file named by the user, but possibly for an in-memory data structure or the result of another operation.